core/src/main/java/org/infinispan/protostream/LazyByteArrayOutputStream.java

```diff
@@ -0,0 +1,27 @@
+package org.infinispan.protostream;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+public interface LazyByteArrayOutputStream {
+   int getPosition();
+   void setPosition(int position);
+   void ensureCapacity(int size);
+   byte[] getRawBuffer();
```

Comment:
Is there any reason we aren't just doing `write(int pos, byte val)` instead? This would allow for other implementations that aren't byte[] backed.

Reply:
The original purpose of the interface was to allow different implementations of `byte[]` backed streams, so exposing `byte[]` directly removes the need for lots of additional `#write` methods calls.

I guess your suggestion would allow the interface to be used by other indexed `OutputStream` implementations though, e.g. one that is ByteBuf based. Would this be beneficial for the client?

Reply:
Yes, this is what I am thinking. I think this should benefit both the client and possibly even the server with some more changes I am thinking about.

Reply:
Interface and implementation updated to no longer expose a `byte[]`. Let me know if we're missing anything you think we'll need for your future ideas.